libqtile/layout/floating.py

```diff
@@ -88,19 +86,41 @@ def __init__(self, float_rules=None, no_reposition_match=None, **config):
         the screen.
         """
         Layout.__init__(self, **config)
+        self.add_defaults(Floating.defaults)
         self.clients = []
         self.focused = None
         self.group = None
-        self.float_rules = float_rules or DEFAULT_FLOAT_RULES
         self.no_reposition_match = no_reposition_match
-        self.add_defaults(Floating.defaults)
+
+        if override_default_float_rules:
+            self.float_rules = defaultdict(set)
+        else:
+            self.float_rules = DEFAULT_FLOAT_RULES
+        # provide support for deprecated `auto_float_types` argument
+        if self.auto_float_types:
+            logger.warning("`auto_float_types` is deprecated. Use `float_rules` for this.")
+            self.float_rules["wmtype"].update(self.auto_float_types)
+        # provide support for the deprecated list-structure of `float_rules`
+        if isinstance(float_rules, list):
+            logger.warning("The list structure of `float_rules` is deprecated. Use a dictionary instead")
+            for rule_dict in float_rules:
+                key, val = tuple(rule_dict.items())[0]
+                self.float_rules[key].add(val)
+        else:
+            for key, vals in float_rules.items():
+                self.float_rules[key].update(vals)
 
     def match(self, win):
-        """Used to default float some windows"""
-        if win.window.get_wm_type() in self.auto_float_types:
+        """Compares a window's properties with the `float_rules`.
+        Returns `True` if window should float.
+        """
+        if (win.name in self.float_rules['wmname'] or
+                win.window.get_wm_window_role() in self.float_rules['role'] or
+                set(win.window.get_wm_class()) & self.float_rules['wmclass'] or
+                win.window.get_wm_type() in self.float_rules['wmtype']):
             return True
-        for rule_dict in self.float_rules:
-            if win.match(**rule_dict):
+        if "fixed_size" in self.float_rules['wmhint']:
```

Comment:
Do I understand correctly that `fixed_size` is just a "special" value that you can give to `float_rules["wmhint"]` to trigger that second check, and not a real hint? If so, this doesn't seem like a good API to me. :(

Reply:
Yes. To check this we need minimum and maximum size in WM_NORMAL_HINTS.

Reply:
If we want to give the user the possibiltiy to set this check, it needs to be `float_rules` somehow. Else we have to hardcode this check.

Reply:
But it is in the spec:
> Windows can indicate that they are non-resizable by setting minheight = maxheight and minwidth = maxwidth in the ICCCM WM_NORMAL_HINTS property. The Window Manager MAY decorate such windows differently. 

https://specifications.freedesktop.org/wm-spec/wm-spec-latest.html#NORESIZE
So it's just implementing the API in my opinion.

Reply:
Yeah, it's great that we handle this scenario, but what I don't like is that we need to use a special value with a special case that can not be guessed.

One solution could be that we allow functions inside float_rules, for example. We could have something like:

```python
DEFAULT_FLOAT_RULES = [
    lambda w: w.window.get_wm_type() in ("dialog", "utility", "toolbar", "splash"),
    lambda w: w.hints["max_width"] == w.hints["min_width"],
]
```

That would also have the advantage of just being an addition to the current API, and not a change, as the structure of the rules would stay the same (a list). We would just have to check if the rule is a dict and if so, use `win.match`, or a callable and if so, call it.

Reply:
I think the `fixed_size` value isn't that special as it is mentioned in the X specification. If we put it in the docs, there is no need for guessing.

But I also like the idea with the lambdas because it would allow the user to set very custom rules.